server/aws-lsp-codewhisperer/src/language-server/utilities/telemetryUtils.ts

```diff
@@ -0,0 +1,99 @@
+    }
+
+    // Compute client-specific suffix
+    // Missing required data fields are replaced with 'UNKNOWN' token
+    // Whitespaces in product.name and platform.name are replaced to '-'
+    if (initializeParams?.initializationOptions?.aws) {
+        const { clientInfo } = initializeParams?.initializationOptions?.aws
+        const { extension } = clientInfo || {}
+
+        if (extension) {
+            items.push(`${extension.name ? format(extension.name) : 'UNKNOWN'}/${extension.version || 'UNKNOWN'}`)
+        }
+
+        if (clientInfo) {
+            items.push(`${clientInfo.name ? format(clientInfo.name) : 'UNKNOWN'}/${clientInfo.version || 'UNKNOWN'}`)
+        }
+
+        if (clientInfo?.clientId) {
+            items.push(`ClientId/${clientInfo?.clientId}`)
+        }
+    } else {
+        // Default to standard InitializeParams.clientInfo if no custom aws.clientInfo is set
+        const { clientInfo } = initializeParams || {}
+        if (clientInfo) {
+            items.push(`${clientInfo.name ? format(clientInfo.name) : 'UNKNOWN'}/${clientInfo.version || 'UNKNOWN'}`)
+        }
+    }
+
+    return items.join(' ')
+}
+
+const IDE_CATEGORY_MAP: { [key: string]: IdeCategory } = {
+    'AmazonQ-For-VSCode': 'VSCODE',
+    'Amazon-Q-For-JetBrains': 'JETBRAINS',
+    'AmazonQ-For-Eclipse': 'ECLIPSE',
+    'AWS-Toolkit-For-VisualStudio': 'VISUAL_STUDIO',
+}
+
+const mapClientNameToIdeCategory = (clientName: string): string | undefined => {
+    return IDE_CATEGORY_MAP[clientName]
+}
+
+// Use InitializeParams.initializationOptions.aws to derive IDE Category from calling client.
+const getIdeCategory = (initializeParams: InitializeParams) => {
+    let ideCategory
+    if (initializeParams.initializationOptions?.aws.clientInfo?.extension?.name) {
+        ideCategory = mapClientNameToIdeCategory(initializeParams.initializationOptions.aws.clientInfo.extension.name)
+    }
+
+    return ideCategory || 'UNKNOWN'
+}
+
+// Map result from https://nodejs.org/api/process.html#process_process_platform to expected Operating system
+const getOperatingSystem = () => {
+    switch (process.platform) {
+        case 'darwin':
+            return 'MAC'
+        case 'win32':
+            return 'WINDOWS'
+        case 'linux':
+            return 'LINUX'
+        default:
+            return 'UNKNOWN'
+    }
+}
+
+export const makeUserContextObject = (initializeParams: InitializeParams, product: string): UserContext | undefined => {
+    const userContext: UserContext = {
+        ideCategory: getIdeCategory(initializeParams),
+        operatingSystem: getOperatingSystem(),
+        product: product,
+        clientId: initializeParams?.initializationOptions?.aws?.clientInfo?.clientId,
+        ideVersion:
+            initializeParams?.initializationOptions?.aws?.clientInfo?.version || initializeParams.clientInfo?.version,
+    }
+
+    if (userContext.ideCategory === 'UNKNOWN' || userContext.operatingSystem === 'UNKNOWN') {
+        return
```

Comment:
`SendTelemetryEvent` won't work for web due to auth support. We'll need to make sure we report CW business telemetry for standalone only.

Reply:
Then, we need to log error, if `ideCategory` or `operatingSystem` is undefined.

Reply:
I leave this logic to be implemented in TelemetryService that we are adding in https://github.com/aws/language-servers/pull/524. This PR just adds utility helper to compute this object